### PR TITLE
Voja

### PR DESCRIPTION
**Pull request by neworderofjamie**

### CODE REVIEW

**Comment by mundya** [comment]
I look forward to extracting profiling from this :/

File: nengo_spinnaker/builder/ensemble.py [hunk 1]
```diff
@@ -226,3 +301,29 @@ def build_neurons_probe(model, probe):
                 probe.target.ensemble.neuron_type.__class__.__name__
             )
         )
+
+
+@Model.probe_builders.register(nengo.connection.LearningRule)
+def build_voja_probe(model, probe):
```
Comment (mundya) [inline nengo_spinnaker/builder/ensemble.py:307]: Could this be `build_learnt_connection_probe` or similar?

File: spinnaker_components/Makefile.common [hunk 2]
```diff
@@ -163,7 +172,7 @@ endif
 	$(LS) $(APP_DIR)/$(APP).aplx
 
 
-# Build the ELF file
+# Build the ELF filBUILDDIRe
```
Comment (mundya) [inline spinnaker_components/Makefile.common:175]: Huh?

File: spinnaker_components/ensemble/ensemble.h [hunk 3]
```diff
@@ -34,57 +34,85 @@
...
-extern input_filter_t g_input;     //!< Input filters and buffers
-extern input_filter_t g_input_inhibitory;     //!< Input filters and buffers
-extern input_filter_t g_input_modulatory;     //!< Input filters and buffers
+// Input filters and buffers for general and inhibitory inputs. Their outputs
+// are summed into accumulators which are used to drive the standard neural input
+extern input_filter_t g_input;
+extern input_filter_t g_input_inhibitory;
+
+// Input filters and buffers for modulatory signals. Their
+// outputs are left seperate for use by learning rules
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble.h]: sp. separate

File: spinnaker_components/ensemble/ensemble.h [hunk 4]
```diff
@@ -34,57 +34,85 @@
...
+extern input_filter_t g_input_modulatory;
+
+// Input filters and buffers for signals to be encoded by learnt encoders.
+// Each output is encoded by a seperate encoder so these are also left seperate
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble.h]: sp. separate

File: spinnaker_components/ensemble/ensemble_filtered_activity.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  if(g_num_activity_filters > 0)
+  {
+    // Allocate memory
+    MALLOC_FAIL_FALSE(g_activity_filter_params,
+                      g_num_activity_filters * sizeof(activity_filter_parameters_t));
+
+    MALLOC_FAIL_FALSE(g_filtered_activities,
+                      g_num_activity_filters * sizeof(value_t*));
+
+    // Copy propogators from region into new array
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble_filtered_activity.c:47]: sp.

File: spinnaker_components/ensemble/ensemble_filtered_activity.c [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+}
+//-----------------------------------------------------------------------------
+void filtered_activity_step()
+{
+  profiler_write_entry(PROFILER_ENTER | PROFILER_TIMER_OUTPUT_FILTER);
+
+  // Loop through filters
+  for(uint f = 0; f < g_num_activity_filters; f++)
+  {
+    // Loop through neurons and apply propogators
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble_filtered_activity.c:75]: sp.

File: spinnaker_components/ensemble/ensemble_pes.c [hunk 7]
```diff
@@ -60,9 +47,47 @@ bool get_pes(address_t address)
...
+{
+  profiler_write_entry(PROFILER_ENTER | PROFILER_FILTERED_PES);
+
+  // Loop through all the learning rules
+  for(uint32_t l = 0; l < g_num_pes_learning_rules; l++)
+  {
+    // Extract filtered error signal vector indexed by learning rule
+    const pes_parameters_t *parameters = &g_pes_learning_rules[l];
+
+    // If this learning rule operates on filtered activity and should, therefore be updated here
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble_pes.c:67]: Say what now?

File: spinnaker_components/ensemble/ensemble_pes.h [hunk 8]
```diff
@@ -53,20 +57,21 @@ static inline void pes_neuron_spiked(uint n)
   // Loop through all the learning rules
   for(uint32_t l = 0; l < g_num_pes_learning_rules; l++)
   {
+    // If this learning rule operates on un-filtered activity and should, therefore be updated here
```
Comment (mundya) [inline spinnaker_components/ensemble/ensemble_pes.h:60]: Huh?

**Comment by mundya** [comment]
This might be going through in person if you've time? Sorry

File: spinnaker_components/Makefile.common (hunk 2 above)
Comment (neworderofjamie) [inline spinnaker_components/Makefile.common:175]: I think that's probably a good old fashioned copy-paste flail :)

File: spinnaker_components/ensemble/ensemble_pes.h (hunk 8 above)
Comment (neworderofjamie) [inline spinnaker_components/ensemble/ensemble_pes.h:60]: Sure, am in all day, probably staring at progress bars a lot!

**Comment by mundya** [comment]
This now includes #43 and #37

**Comment by mundya** [comment]
So this is waiting on ensuring that everything still runs in PES and Voja are not present/the expected version?

**Comment by mundya** [comment]
Closing this PR because the long suffering @neworderofjamie is working on another implementation which builds on #87.

**Pull request closed without merging**